kotlinx-coroutines-core/common/src/flow/operators/Delay.kt

```diff
@@ -345,3 +346,105 @@ internal fun CoroutineScope.fixedPeriodTicker(delayMillis: Long, initialDelayMil
  */
 @FlowPreview
 public fun <T> Flow<T>.sample(period: Duration): Flow<T> = sample(period.toDelayMillis())
+
+/**
+ * Returns a flow that will emit a [FlowTimeoutException] if the upstream doesn't emit an item within the given time.
+ *
+ * Example:
+ *
+ * ```kotlin
+ * flow {
+ *     emit(1)
+ *     delay(100)
+ *     emit(2)
+ *     delay(100)
+ *     emit(3)
+ *     delay(1000)
+ *     emit(4)
+ * }.timeout(100.milliseconds).catch {
+ *     emit(-1) // Item to emit on timeout
+ * }.onEach {
+ *     delay(300) // This will not cause a timeout
+ * }
+ * ```
+ * <!--- KNIT example-timeout-duration-01.kt -->
+ *
+ * produces the following emissions
+ *
+ * ```text
+ * 1, 2, 3, -1
+ * ```
+ * <!--- TEST -->
+ *
+ * Note that delaying on the downstream doesn't trigger the timeout.
+ *
+ * @param timeout Timeout period
+ */
+@FlowPreview
+public fun <T> Flow<T>.timeout(
+    timeout: Duration
+): Flow<T> = timeoutInternal(timeout.toDelayMillis())
+
+private fun <T> Flow<T>.timeoutInternal(
+    timeoutMillis: Long
+): Flow<T> = scopedFlow { downStream ->
+    if (timeoutMillis <= 0L) throw FlowTimeoutException("Timed out immediately")
+
+    // Produce the values using the default (rendezvous) channel
+    // Similar to [debounceInternal]
+    val values = produce<Any?> {
```

Comment:
Also, it seems like there exists both more straightforward, idiomatic and efficient solution (tho I haven't really tested it):

```
private fun <T> Flow<T>.timeoutInternal(
    timeoutMillis: Long
): Flow<T> = scopedFlow { downStream ->
    if (timeoutMillis <= 0L) throw FlowTimeoutException("Timed out immediately")
    val values = buffer(Channel.RENDEZVOUS).produceIn(this)
    whileSelect {
        values.onReceiveCatching { value ->
            value.onSuccess {
                downStream.emit(it)
            }.onClosed {
                return@onReceiveCatching false
            }
            return@onReceiveCatching true
        }
        onTimeout(timeoutMillis) {
            throw FlowTimeoutException(timeoutMillis)
        }
    }
}
```

Could you please take a look at it?

Reply:
I'm seeing `testScalar` and `testEmpty` tests failing intermittently with this implementation. It seems that it takes some measure of time to hit `onReceiveCatching`, which is just long enough in the case of these two tests to trigger the `onTimeout()` select block.

Not entirely sure why this fails, as I believe it is very close to what I also had, but my guess is that the `invokeOnTimeout()` in the `Select.kt` is posted on a different context or dispatcher, rather than the same one used for the test (given these tests are running with `runTest` instead of `withVirtualTime`).

Reply:
I'll check. `withVirtualTime` is really hacky and written in an an ad-hoc manner, so maybe it causes unforeseen issues

Reply:
I don't think the problem is with `withVIrtualTime` in this case, as both tests are running in `runTest`. I dug a bit deeper, and saw that the `onTimeout` was eventually leading to `EventLoopImplBase.scheduleInvokeOnTimeout`, which led to `DelayedRunnableTask`, where in my implementation, it would go to `EventLoopImplBase.scheduleResumeAfterDelay` which leads to the `DelayedResumeTask`. I believe it's this difference that's causing the tests to be flaky.

Reply:
Please update your branch on the most recent `develop`, and then this implementation should pass all the tests.

I suggest using it as the most straightforward and performant

Reply:
(just to be clear, "this implementation" -- the one proposed in the comment https://github.com/Kotlin/kotlinx.coroutines/pull/3148#discussion_r981143278)

Reply:
Although the tests are much more stable now, there is still a bit of flakiness with this implementation for the same two tests (`testScalar` and `testEmpty`). However, it is still much more performant, and I have to repeat the tests a few million times to even see a failure. I'll make the change, but add a comment to both of these tests about them being flaky.